Fix #144

to266 opened the pull request:
Fixes #144 
Uses the first suggested solution by propagating the network error through the connection.

~~Will test in actual library that uses eventually-rs in a couple of hours, to see if this intrinsically solves the issue~~ Yes works

ar3s3ru commented on eventually-postgres/src/subscriber.rs:
```diff
@@ -23,20 +23,26 @@ use tokio_postgres::AsyncMessage;
+    /// Error variant returned when the connection, used for `LISTEN` asynchronous notifications
+    /// gets dropped. Currently the subscriber cannot recover from this error and a new one should
+    /// be created.
+    #[error("postgres connection error: ${0}")]
```
```suggestion
    #[error("postgres connection error: {0}")]
```
Why using `${0}`?

ar3s3ru commented on eventually-postgres/src/subscriber.rs:
```diff
@@ -23,20 +23,26 @@ use tokio_postgres::AsyncMessage;
+    /// gets dropped. Currently the subscriber cannot recover from this error and a new one should
+    /// be created.
+    #[error("postgres connection error: ${0}")]
+    ConnectionError(String),
```
Can we use `tokio_postgres::Error` here instead of `String`?

```suggestion
    ConnectionError(#[source] tokio_postgres),
```

to266 replied:
`tokio_postgres` error is not `Clone`, and it needs to be `Clone` to send over the channel, as far as I can tell

ar3s3ru replied:
Oh yes, indeed. Another "smell" of using a channel to send these values around 😅

ar3s3ru commented on eventually-postgres/src/subscriber.rs:
```diff
@@ -23,20 +23,26 @@ use tokio_postgres::AsyncMessage;
-#[error("failed to deserialize notification payload from JSON: {message}")]
-pub struct DeserializeError {
-    message: String,
+pub enum SubscriberError {
```
If we have `SubscriberError`, we can remove the duplicated "Error" part from the enum variants:

```rust
pub enum SubscriberError {
    Deserialize(String),
    Connection(tokio_postgres::Error),
}
```

ar3s3ru reviewed (commented):
Thanks for the PR :heart: 

Although I don't think this should be the end design, as I wrote on #144, we can merge this PR (when comments are addressed) to fix the bug. However, if you're still interested, I'd like to discuss and expand a better design approach on the original issue :smile:

to266 commented:
Addressed the comments :smiley: 

As far as #144 suggestion goes, I'm a fan of that! Right now it's a bit strange, as the background process starts when creating the subscriber, and until `subscribe_all` is invoked, I guess the channel simply fills with stuff... Moving this to `subscribe_all` makes it way more intuitive :+1: 

The only issue from my side right now is the lack of time for moving things along :wink:

ar3s3ru commented:
@to266 glad to hear that! I will create a new issue (or use the new Discussions space) to keep digging at the design :+1:
